### DISCUSSION TITLE
Dependency injection cater class

### DISCUSSION THREAD

**Discussion by Eroha186** [laravel/framework#53900] [Q&A]
Hi everyone,

I'm trying to use dependency injection in the constructor of a custom cast in my Laravel application. However, it seems like the usual approach for injecting dependencies isn't working here.
For example:
```php
class CustomCast implements CastsAttributes
{
    public function __construct(protected SomeService $service)
    {
    }

    public function get($model, $key, $value, $attributes)
    {
        // Some logic using $this->service
    }

    public function set($model, $key, $value, $attributes)
    {
        // Some logic using $this->service
    }
}
```

When I attempt this, Laravel throws an error, as it doesn't seem to resolve the dependencies automatically.

Are there workarounds or best practices for achieving this in custom casts without resorting to calling the container manually using the `app()` helper function?

Thanks in advance for any insights!

**Comment by macropay-solutions** [comment]
Hi,
You are declaring twice the property in your code example, the 2nd time in the construct definition.
Is laravel supposed to offer this autowirring functionality on cast Classes?

Update. 
Have you tried it like:

```php
class CustomCast implements CastsAttributes
{
    protected SomeService $service;

    public function __construct(SomeService $service)
        $this->service = $service;
    {
    }

    public function get($model, $key, $value, $attributes)
    {
        // Some logic using $this->service
    }

    public function set($model, $key, $value, $attributes)
    {
        // Some logic using $this->service
    }
}

```

**Reply by Eroha186** [reply]
Sorry for the incorrect example. I will update my code example.
Your code fragment is correct, but mine is inaccurate. 
Yes, I wanted to know if it is possible to make autowirring or if this functionality will be added in the future

**Comment by henzeb** [comment]
Not sure if casts are resolved using the service container, but you can resolve it using the `resolve()` helper instead?

**Reply by Eroha186** [reply]
Thanks, you're right, I can reuse the helper `resolve()`, but it generates a composition that I'd like to avoid in my case study

**Comment by Eroha186** [comment]
I found the answer to my question and casts do not allow dependency injection in the constructor this is seen in [trait HasAttributes](https://github.com/laravel/framework/blob/11.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L1788)

Does anyone have any ideas on how to allow Laravel to do dependency injection in casts, it seems convenient to me and this behaviour would be expected as in other parts of my application I can freely use the dependency injection container in the constructor

**Reply by henzeb** [reply]
The best way? You can propose a PR with a change. 
The semi best way? override the resolveCasterClass() method.

**Reply by Eroha186** [reply]
Thanks for the reply!
Will this be needed by the community or is this my isolated problem

**Reply by ghost** [reply]
I see it as a no problem. This is why you have resolve or app. (side note as a joke, have a look here https://github.com/laravel/framework/discussions/53917)

**Reply by Eroha186** [reply]
I see the problem here is that I expected dependency injection when I wrote the constructor for my caster, as it is done in other parts of the framework.
That's a great joke!)

**Reply by Sophist-UK** [reply]
Dependency injection is a widely adopted pattern in Laravel, and I am not sure why casts don't have it.

But equally I am unsure what the use case is - dependency injection is supposed to allow you flexibility to replace the dependent service with e.g. a mock service for testing, so is that the use case here?

If you want to allow dependency injection, then I think you would need to ensure backwards compatibility by making the existing dependency a default if it is not specified.

**Reply by Eroha186** [reply]
Yes, you are right, when writing unit tests I can clearly see what services I need for the class to work, which I think makes the code more readable and I can freely write the necessary mock without diving into the class implementation.

Also the ability to enforce dependencies in the caster will add predictability to the coding behavior, because as you mentioned earlier, dependency enforcement is a common patter in Laravel.
